### DISCUSSION TITLE
Implement a new storage type for non-bonded pair interactions

### DISCUSSION THREAD

**Discussion by KaiSzuttor** [espressomd/espresso#4216] [Ideas]
Idea: use the Cantor pairing bijective function to map from type pair to the index of the pair interaction in a vector of tuples of interactions.

```c++

 std::size_t cantor_pairing(std::pair<std::size_t, std::size_t> const &p) const {
   auto hash = [](std::size_t k1, std::size_t k2){ return (k1 + k2) / 2 * (k1 + k2 + 1) + k2; };
   // make it symmetric
   if (p.second > p.first) {
     return hash(p.first, p.second);
   }
   return hash(p.second, p.first);
 }



class NonBondedRegistry {
  std::vector<std::tuple<LennardJones, Gaussian, Hat>> m_storage;

  template<class NonBondedInteraction>
  void add_interaction(std::pair<std::size_t, std::size_t> const &types, NonBondedInteraction inter) {
    auto const index = cantor_pairing(types);
    if (m_storage.size() < index) m_storage.resize(index);
    std::get<NonBondedInteraction>(m_storage[index]) = inter;
  }
};

**Comment by fweik** [comment]
Can you explain why cantor pairing? I think the current hash function is better suited for the typical use case, due to better locality.

Currently the implementation looks like this:

```c++
inline IA_parameters *get_ia_param(int i, int j) {
  return &ia_params[Utils::upper_triangular(std::min(i, j), std::max(i, j),
                                            max_seen_particle_type)];
}
```

**Reply by KaiSzuttor** [reply]
Just because of simplicity... Can you explain what you exactly mean with "better locality"?

**Reply by fweik** [reply]
I mean that the interaction parameters for one (left) type are consecutive in memory (that is one row in the upper triangular matrix), which I now realize is not completely true because of the symmetrization, but still. With the cantor pairing you are basically guaranteed to jump around.

**Reply by KaiSzuttor** [reply]
okay, so keeping the current hashing would of course also be possible.

**Reply by fweik** [reply]
Ideally one would implement a hash table for perfect hash functions, and then do benchmarks with different hash functions.

**Reply by KaiSzuttor** [reply]
This may also be of interest, it discusses the point you raised in greater detail:

https://www.worldscientific.com/doi/pdf/10.1142/S012905410300156X
